### PR TITLE
Add missing type hints to synapse.replication.

### PR DESCRIPTION
**Pull request by clokep**
Follow-up to #11856 which completes the type hints in `synapse.replication`.

I think this was all pretty straightforward?

### CODE REVIEW

File: synapse/replication/tcp/streams/events.py [hunk 1]
```diff
@@ -14,11 +14,16 @@
 # limitations under the License.
 import heapq
 from collections.abc import Iterable
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/replication/tcp/streams/events.py:16]: c.f. https://stackoverflow.com/a/59956132/5252017

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/replication/tcp/streams/events.py:16]: Yep, this isn't ready for review yet.

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/replication/tcp/streams/events.py]: Seems that fixed it though! Thanks for the hint. I fixed another instance too while I was at it.

File: synapse/replication/tcp/commands.py [hunk 2]
```diff
@@ -109,14 +115,16 @@ class RdataCommand(Command):
...
+        self, stream_name: str, instance_name: str, token: Optional[int], row: StreamRow
+    ):
         self.stream_name = stream_name
         self.instance_name = instance_name
         self.token = token
         self.row = row
 
     @classmethod
-    def from_line(cls, line):
+    def from_line(cls: Type["RdataCommand"], line: str) -> "RdataCommand":
```
Comment (DMRobertson) [inline synapse/replication/tcp/commands.py:127]: Just for info: [`typing.Self`](https://www.python.org/dev/peps/pep-0673/#use-in-classmethod-signatures) might make this easier in the future. I think it's in new/bleeding edge `typing_extensions` but probably not something we want to opt into for now.

File: tests/replication/_base.py [hunk 3]
```diff
@@ -53,7 +54,7 @@ def prepare(self, reactor, clock, hs):
         server_factory = ReplicationStreamProtocolFactory(hs)
         self.streamer = hs.get_replication_streamer()
         self.server: ServerReplicationStreamProtocol = server_factory.buildProtocol(
-            None
+            IPv4Address("TCP", "127.0.0.1", 0)
```
Comment (DMRobertson) [inline tests/replication/_base.py:57]: Aside: slightly surprised to see TCP being part of an IP address struct. Isn't TCP a layer up from IP? /shrug

**Review by DMRobertson — APPROVED** [review]
LGTM, a few thoughts though:

File: synapse/replication/tcp/commands.py (hunk 2 above)
Comment (clokep) [reply synapse/replication/tcp/commands.py:127]: Ah, had not seen that before! I was following a pattern we use in a few other spots too.

File: tests/replication/_base.py (hunk 3 above)
Comment (clokep) [reply tests/replication/_base.py:57]: Twisted considers addresses to be many different things: IP, pipes, files, etc.

File: synapse/replication/slave/storage/devices.py [hunk 4]
```diff
@@ -60,7 +60,9 @@ def __init__(
     def get_device_stream_token(self) -> int:
         return self._device_list_id_gen.get_current_token()
 
-    def process_replication_rows(self, stream_name, instance_name, token, rows):
+    def process_replication_rows(
+        self, stream_name: str, instance_name: str, token: int, rows: Iterable[Any]
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/replication/slave/storage/devices.py:64]: Maybe there's some way to use Generics, but... not one for now.

[the remark this answers is not in the document]
Comment (clokep) [reply synapse/replication/slave/storage/devices.py:64]: I think it actually has to accept them all? It is a bit annoying though cause the (`stream_name`, `row_type`) is essentially paired, but I don't think we can tell mypy that? Essentially `DeviceListsStream.DeviceListsStreamRow` only ever occurs when we have `stream_name = 'device_stream'`...

Anyway, yes I think this is not one for now. 👍

**Pull request merged**